File: app/src/main/java/com/ywwynm/everythingdone/activities/HelpActivity.java
package com.ywwynm.everythingdone.activities;

import android.support.v4.app.FragmentTransaction;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.ywwynm.everythingdone.R;
import com.ywwynm.everythingdone.adapters.BaseViewHolder;
import com.ywwynm.everythingdone.fragments.HelpDetailFragment;
import com.ywwynm.everythingdone.helpers.SendInfoHelper;
import com.ywwynm.everythingdone.utils.DisplayUtil;
import com.ywwynm.everythingdone.utils.EdgeEffectUtil;

public class HelpActivity extends EverythingDoneBaseActivity {

    private String[] mTitles;
    private String[] mContents;

    private HelpDetailFragment mHelpDetailFragment;

    private RecyclerView mRecyclerView;

    @Override
    protected int getLayoutResource() {
        return R.layout.activity_help;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_help, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.act_feedback) {
            SendInfoHelper.sendFeedback(this, false);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void initMembers() {
        mTitles   = getResources().getStringArray(R.array.help_titles);
        mContents = getResources().getStringArray(R.array.help_contents);
    }

    @Override
    protected void findViews() {

    }

    @Override
    protected void initUI() {
        DisplayUtil.expandLayoutToStatusBarAboveLollipop(this);
        DisplayUtil.expandStatusBarViewAboveKitkat(f(R.id.view_status_bar));
        DisplayUtil.darkStatusBar(this);

        mRecyclerView = f(R.id.rv_help);
        mRecyclerView.setAdapter(new HelpAdapter());
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));

        mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            final int color = ContextCompat.getColor(HelpActivity.this, R.color.blue_deep);
            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                EdgeEffectUtil.forRecyclerView(mRecyclerView, color);
            }
        });
    }

    @Override
    protected void setActionbar() {
        final Toolbar toolbar = f(R.id.actionbar);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mHelpDetailFragment != null && mHelpDetailFragment.isVisible()) {
                    getSupportFragmentManager().popBackStack();
                } else {
                    finish();
                }
            }
        });
    }

    @Override
    protected void setEvents() {

    }

    public void updateActionBarTitle(boolean toDetail) {
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setTitle(toDetail ? R.string.help_detail : R.string.help);
        }
    }

    // Used to make the RecyclerView not focusable in talkback mode
    public void setRecyclerViewFocusable(boolean focusable) {
        if (focusable) {
            mRecyclerView.setVisibility(View.VISIBLE);
        } else {
            mRecyclerView.setVisibility(View.GONE);
        }
    }

    class HelpAdapter extends RecyclerView.Adapter<HelpAdapter.HelperHolder> {

        private LayoutInflater mInflater = LayoutInflater.from(HelpActivity.this);

        @Override
        public HelperHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            return new HelperHolder(mInflater.inflate(R.layout.rv_help, parent, false));
        }

        @Override
        public void onBindViewHolder(HelperHolder holder, int position) {
            holder.tv.setText(mTitles[position]);
        }

        @Override
        public int getItemCount() {
            return mTitles.length;
        }

        class HelperHolder extends BaseViewHolder {

            TextView tv;

            HelperHolder(View itemView) {
                super(itemView);

                tv = f(R.id.tv_help_rv);

                f(R.id.ll_help_rv).setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        final int pos = getAdapterPosition();
                        mHelpDetailFragment = HelpDetailFragment.newInstance(
                                mTitles, mContents, pos);
                        final String tag = HelpDetailFragment.TAG;
                        getSupportFragmentManager()
                                .beginTransaction()
                                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                                .replace(R.id.fl_fragment_container_help,
                                        mHelpDetailFragment, tag)
                                .addToBackStack(tag)
                                .commit();
                    }
                });
            }
        }

    }

}

File: app/src/main/java/com/ywwynm/everythingdone/activities/HelpActivity.kt
@file:Suppress("DEPRECATION")

package com.ywwynm.everythingdone.activities

import androidx.fragment.app.FragmentTransaction
import androidx.core.content.ContextCompat
import androidx.appcompat.app.ActionBar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.appcompat.widget.Toolbar
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.ywwynm.everythingdone.R
import com.ywwynm.everythingdone.adapters.BaseViewHolder
import com.ywwynm.everythingdone.fragments.HelpDetailFragment
import com.ywwynm.everythingdone.helpers.SendInfoHelper
import com.ywwynm.everythingdone.utils.DisplayUtil
import com.ywwynm.everythingdone.utils.EdgeEffectUtil

open class HelpActivity : EverythingDoneBaseActivity() {

    private var mTitles: Array<String?>? = null
    private var mContents: Array<String?>? = null

    private var mHelpDetailFragment: HelpDetailFragment? = null

    private var mRecyclerView: RecyclerView? = null

    override fun getLayoutResource(): Int = R.layout.activity_help

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_help, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.act_feedback) {
            SendInfoHelper.sendFeedback(this, false)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun initMembers() {
        mTitles   = resources.getStringArray(R.array.help_titles) as Array<String?>
        mContents = resources.getStringArray(R.array.help_contents) as Array<String?>
    }

    override fun findViews() {
    }

    override fun initUI() {
        DisplayUtil.expandLayoutToStatusBarAboveLollipop(this)
        DisplayUtil.expandStatusBarViewAboveKitkat(f(R.id.view_status_bar))
        DisplayUtil.darkStatusBar(this)

        mRecyclerView = f(R.id.rv_help)
        mRecyclerView!!.adapter = HelpAdapter()
        mRecyclerView!!.layoutManager = LinearLayoutManager(this)
        DisplayUtil.applyBottomInsetAsScrollPadding(mRecyclerView)

        mRecyclerView!!.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            val color: Int = ContextCompat.getColor(this@HelpActivity, R.color.blue_deep)
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                EdgeEffectUtil.forRecyclerView(mRecyclerView, color)
            }
        })
    }

    override fun setActionbar() {
        val toolbar: Toolbar = f(R.id.actionbar)!!
        setSupportActionBar(toolbar)
        val actionBar: ActionBar? = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)
        toolbar.setNavigationOnClickListener {
            if (mHelpDetailFragment != null && mHelpDetailFragment!!.isVisible) {
                supportFragmentManager.popBackStack()
            } else {
                finish()
            }
        }
    }

    override fun setEvents() {
    }

    open fun updateActionBarTitle(toDetail: Boolean) {
        val actionBar: ActionBar? = supportActionBar
        actionBar?.setTitle(if (toDetail) R.string.help_detail else R.string.help)
    }

    // Used to make the RecyclerView not focusable in talkback mode
    open fun setRecyclerViewFocusable(focusable: Boolean) {
        if (focusable) {
            mRecyclerView!!.visibility = View.VISIBLE
        } else {
            mRecyclerView!!.visibility = View.GONE
        }
    }

    internal inner class HelpAdapter : RecyclerView.Adapter<HelpAdapter.HelperHolder>() {

        private val mInflater: LayoutInflater = LayoutInflater.from(this@HelpActivity)

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HelperHolder {
            return HelperHolder(mInflater.inflate(R.layout.rv_help, parent, false))
        }

        override fun onBindViewHolder(holder: HelperHolder, position: Int) {
            holder.tv!!.text = mTitles!![position]
        }

        override fun getItemCount(): Int = mTitles!!.size

        internal inner class HelperHolder(itemView: View?) : BaseViewHolder(itemView) {

            var tv: TextView? = f(R.id.tv_help_rv)

            init {
                f<View>(R.id.ll_help_rv)!!.setOnClickListener {
                    val pos = adapterPosition
                    mHelpDetailFragment = HelpDetailFragment.newInstance(
                        mTitles, mContents, pos
                    )
                    val tag = HelpDetailFragment.TAG
                    supportFragmentManager
                        .beginTransaction()
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                        .replace(R.id.fl_fragment_container_help,
                            mHelpDetailFragment!!, tag)
                        .addToBackStack(tag)
                        .commit()
                }
            }
        }
    }
}
